src/themis/secure_message_wrapper.c

```diff
@@ -300,6 +336,15 @@ themis_status_t themis_secure_message_rsa_decrypter_proceed(themis_secure_messag
                        == THEMIS_SECURE_MESSAGE_RSA_ENCRYPTED);
     THEMIS_CHECK_PARAM(((const themis_secure_encrypted_message_hdr_t*)wrapped_message)->message_hdr.message_length
                        == wrapped_message_length);
+    /*
+     * Make sure the code below does not trigger an underflow if the header is corrupted.
+     * The subtraction subexpression does not underflow because of the check we made before.
+     * (And yes, this code needs cleanup. I intentionally leave it ugly.)
+     */
+    if ((wrapped_message_length - sizeof(themis_secure_rsa_encrypted_message_hdr_t))
+        < ((const themis_secure_rsa_encrypted_message_hdr_t*)wrapped_message)->encrypted_passwd_length) {
+        return THEMIS_FAIL;
+    }
```

Comment:
So much complex. What is happening here?
Based on these few lines, my guess is that encrypted password should not be larger than encrypted message w/o header. But... I still got some questions

Reply:
> So much complex. What is happening here?

I have absolutely no clue 😝

I just want to make sure *this* does not cause a buffer overrun:

https://github.com/cossacklabs/themis/blob/647fb0cc8c3939ec17a5fd4e89795846358c5eee/src/themis/secure_message_wrapper.c#L309-L311

*and this* does not cause an integer overflow:

https://github.com/cossacklabs/themis/blob/647fb0cc8c3939ec17a5fd4e89795846358c5eee/src/themis/secure_message_wrapper.c#L312-L315

I don't have the details of Secure Message implementation in my head right now, but IIRC Secure Messages encrypted with RSA keys contain an ephemeral symmetric key in them – the thing *actually* encrypted with RSA – and the message itself is encrypted with that ephemeral key for performance and security reasons (RSA is slow and can't handle big blocks securely).